kani-driver/src/exe_trace.rs

```diff
@@ -0,0 +1,281 @@
+    vec![det_val_2], ...
+    */
+    let vecs_as_str = det_vals
+        .iter()
+        .zip(interp_det_vals.iter())
+        .map(|(det_val, interp_det_val)| format!("// {interp_det_val}\nvec!{:?}", det_val))
+        .collect::<Vec<String>>()
+        .join(",\n");
+
+    // Hash the generated det val string along with the proof harness name.
+    let mut hasher = DefaultHasher::new();
+    harness_name.hash(&mut hasher);
+    vecs_as_str.hash(&mut hasher);
+    let hash = hasher.finish();
+
+    let exe_trace_func_name = format!("kani_exe_trace_{harness_name}_{hash}");
+    let exe_trace = format!(
+        "
+        #[test]
+        fn {exe_trace_func_name}() {{
+            let det_vals: Vec<Vec<u8>> = vec![{vecs_as_str}];
+            kani::exe_trace_init(det_vals);
+            {harness_name}();
+        }}"
+    );
+
+    (exe_trace, exe_trace_func_name)
+}
+struct FileLineRange {
+    file: String,
+    line_range: Option<(usize, usize)>,
+}
+
+/// Read the CBMC output, parse it as a JSON object, and extract the deterministic values.
+mod parser {
+    use crate::util::append_path;
+    use serde_json::Value;
+    use std::fs::File;
+    use std::io::BufReader;
+    use std::path::Path;
+
+    /// Extract deterministic values from a failing harness.
+    pub fn get_det_vals(file: &Path) -> (Vec<Vec<u8>>, Vec<String>) {
+        let output_filename = append_path(file, "cbmc_output");
+        let cbmc_out = get_cbmc_out(&output_filename);
+        handle_cbmc_out(&cbmc_out)
+    }
+
+    /// Read in the CBMC results file and deserialize it to a JSON object.
+    fn get_cbmc_out(results_filename: &Path) -> Value {
+        let results_file = File::open(results_filename).unwrap();
+        let reader = BufReader::new(results_file);
+        serde_json::from_reader(reader).unwrap()
+    }
+
+    /// The first-level JSON object parser. This extracts the result message.
+    fn handle_cbmc_out(cbmc_out: &Value) -> (Vec<Vec<u8>>, Vec<String>) {
+        let mut det_vals: Vec<Vec<u8>> = Vec::new();
+        let mut interp_det_vals: Vec<String> = Vec::new();
+        for general_msg in cbmc_out.as_array().unwrap() {
+            let result_msg = &general_msg["result"];
+            if !result_msg.is_null() {
+                for result_val in result_msg.as_array().unwrap() {
+                    handle_result(result_val, &mut det_vals, &mut interp_det_vals);
+                }
+            }
+        }
+        (det_vals, interp_det_vals)
+    }
+
+    /// The second-level JSON object parser. This extracts the traces of failing assertions.
+    fn handle_result(
+        result_val: &Value,
+        det_vals: &mut Vec<Vec<u8>>,
+        interp_det_vals: &mut Vec<String>,
+    ) {
+        let desc = result_val["description"].to_string();
+        let status = result_val["status"].to_string();
+
+        if desc.contains("assertion failed") && status == "\"FAILURE\"" {
+            for trace_pt in result_val["trace"].as_array().unwrap() {
```

Comment:
Heads-up: In #1433 I'm defining structs that can be directly deserialized from these JSON objects (some trace components are available). But we can wait until that's merged to simplify this code.

Reply:
Yeah, this code is a bit hard to read. But I guess we can leave it for a follow up PR.

Reply:
Are there things I can do to make this code easier to read?

Reply:
TODO: Add comment to change this parser once Adrian's PR is merged.

Reply:
Done